README.md
# Twinkly Twinkly Little Star

`ttls` is a small package to help you make async requests to Twinkly LEDs. A command line tool (also called `ttls`) is also included, as well as some examples how to create both loadable movies and realtime sequences.

Written based on the [excellent XLED documentation](https://xled-docs.readthedocs.io/en/latest/) by [@scrool](https://github.com/scrool).


CHANGES.md
# Release Notes

## 1.6.1 (2022-12-07)

- Fix spelling inconsistency in function name

## 1.6.0 (2022-12-06)

- Add handling of predefined effects and playlists (from @Olen)
- Get current color (from @Olen)

## 1.5.1 (2022-11-28)

- Relax `aiohttp` dependency

## 1.5.0 (2022-11-28)

- Get saved and current movies (from @Olen)
- Actually set a static colour (from @Olen), old function available as `set_cycle_colours`
- Support setting default mode for `turn_on` (suggested @Mr-Groch)
- Require Python 3.9

## 1.4.2 (2021-11-24)

- Fix first reponse after authentication failures
- Tweek log levels

## 1.4.1 (2021-11-24)

- Export `host` attribute

## 1.4.0 (2021-11-24)

- Support shared `ClientSession`
- Lower timeout (default 3 seconds)

## 1.3.0 (2021-11-22)

- Add support for setting brightness (from @MizterB)
- Multiple colours in static movie (from @MizterB)

## 1.2.0 (2020-12-21)

- Add support from Twinkly Music (from @MizterB)
- Improve authentication handling (from @MizterB)

## 1.1.0 (2019-12-02)

- Add more movie parameters
- Better command parsing in CLI

## 1.0.0 (2019-11-30)

- Initial release


tests/test_config.py
import unittest

from ttls.client import Twinkly


class TestTwinklyConfig(unittest.TestCase):
    def setUp(self):
        self.client = Twinkly(host="192.0.2.1")

    def test_set_default_mode(self):
        self.client.default_mode = "movie"
        self.assertEqual(self.client.default_mode, "movie")
        with self.assertRaises(ValueError):
            self.client.default_mode = "b0rken"


tests/test_colours.py
import unittest
from typing import Dict, List, Optional, Tuple, Union

from ttls.colours import TwinklyColour


class TestTwinklyColours(unittest.TestCase):
    def test_colours_rgb(self):
        col = TwinklyColour(1, 2, 3)
        self.assertEqual(col.as_twinkly_tuple(), (1, 2, 3))
        self.assertEqual(tuple(col), (1, 2, 3))
        self.assertEqual(col.as_dict(), {"blue": 3, "green": 2, "red": 1})

        col = TwinklyColour.from_twinkly_tuple((1, 2, 3))
        self.assertEqual(col.as_twinkly_tuple(), (1, 2, 3))
        self.assertEqual(tuple(col), (1, 2, 3))
        self.assertEqual(col.as_dict(), {"blue": 3, "green": 2, "red": 1})

        rgb = TwinklyColour.from_twinkly_tuple((1, 2, 3))
        self.assertEqual(rgb.as_dict(), {"blue": 3, "green": 2, "red": 1})

    def test_colours_rgbw(self):
        col = TwinklyColour(1, 2, 3, 4)
        self.assertEqual(col.as_twinkly_tuple(), (4, 1, 2, 3))
        self.assertEqual(tuple(col), (1, 2, 3, 4))
        self.assertEqual(col.as_dict(), {"blue": 3, "green": 2, "red": 1, "white": 4})

        rgbw = TwinklyColour.from_twinkly_tuple((1, 2, 3, 4))
        self.assertEqual(rgbw.as_dict(), {"blue": 4, "green": 3, "red": 2, "white": 1})


tests/test_generic.py
import logging
import unittest
import uuid
from typing import Any

import aiounittest

from ttls.client import Twinkly, TwinklyFrame

_LOGGER = logging.getLogger(__name__)


class TwinklyMock(Twinkly):
    async def send_frame(self, frame: TwinklyFrame) -> None:
        _LOGGER.debug("MOCK: send_frame()")

    async def _post(self, endpoint: str, **kwargs) -> Any:
        _LOGGER.debug("MOCK: POST endpoint %s", endpoint)

    async def _get(self, endpoint: str, **kwargs) -> Any:
        _LOGGER.debug("MOCK: GET endpoint %s", endpoint)
        if endpoint == "gestalt":
            return {
                "product_name": "Twinkly",
                "hardware_version": "100",
                "bytes_per_led": 3,
                "hw_id": "e00000",
                "flash_size": 64,
                "led_type": 14,
                "product_code": "TWS250STP-B",
                "fw_family": "F",
                "device_name": "Xmas tree",
                "uptime": "21172191",
                "mac": "aa:bb:cc:dd:ee:ff",
                "uuid": str(uuid.uuid4()),
                "max_supported_led": 500,
                "number_of_led": 250,
                "led_profile": "RGB",
                "frame_rate": 23.77,
                "measured_frame_rate": 25,
                "movie_capacity": 5397,
                "max_movies": 55,
                "copyright": "LEDWORKS 2021",
                "code": 1000,
            }
        _LOGGER.warning("Endpoint %s not yet implemented")
        return


class TestTwinklyGeneric(aiounittest.AsyncTestCase):
    @classmethod
    def setUpClass(self):
        logging.basicConfig(level=logging.DEBUG)

    def setUp(self):
        self.client = TwinklyMock(host="192.0.2.1")

    async def test_get_details(self):
        res = await self.client.get_details()
        self.assertEqual(res["product_name"], "Twinkly")


if __name__ == "__main__":
    unittest.main()


ttls.py
"""
Twinkly Twinkly Little Star
https://github.com/jschlyter/ttls

Copyright (c) 2019 Jakob Schlyter. All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
"""

import argparse
import base64
import json
import logging
import os
import socket
import time
from typing import List, Tuple

import requests

logger = logging.getLogger(__name__)

TWINKLY_MODES = ['rt', 'movie', 'off', 'demo', 'effect']
TWINKLY_FRAME = List[Tuple[int, int, int]]


class Twinkly(object):

    def __init__(self, host: str, login: bool = True):
        self.logger = logging.getLogger(self.__class__.__name__)
        self.session = requests.Session()
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.host = host
        self.rt_port = 7777
        self.expires = None
        self._token = None
        self._length = None
        if login:
            self.ensure_token()

    @property
    def base(self) -> str:
        return f"http://{self.host}/xled/v1"

    def _post(self, endpoint, **kwargs):
        self.ensure_token()
        self.logger.info("POST endpoint %s", endpoint)
        if 'json' in kwargs:
            self.logger.info("POST payload %s", kwargs['json'])
        return self.session.post(f"{self.base}/{endpoint}", **kwargs)

    def _get(self, endpoint, **kwargs):
        self.ensure_token()
        self.logger.info("GET endpoint %s", endpoint)
        return self.session.get(f"{self.base}/{endpoint}", **kwargs)

    def ensure_token(self):
        if self.expires is None or self.expires <= time.time():
            self.logger.debug("Authentication token expired, will refresh")
            self.login()
            self.verify_login()
        else:
            self.logger.debug("Authentication token still valid")

    def login(self):
        challenge = base64.b64encode(os.urandom(32)).decode()
        payload = {"challenge": challenge}
        response = self.session.post(f"{self.base}/login", json=payload)
        response.raise_for_status()
        r = response.json()
        self._token = r['authentication_token']
        self.session.headers['X-Auth-Token'] = self._token
        self.expires = time.time() + r['authentication_token_expires_in']

    def logout(self):
        response = self._post('logout', json={})
        response.raise_for_status()
        self._token = None

    def verify_login(self):
        response = self._post('verify', json={})
        response.raise_for_status()

    @property
    def token(self) -> str:
        self.ensure_token()
        return self._token

    @property
    def length(self) -> int:
        if self._length is None:
            self._length = self.get_details()['number_of_led']
        return self._length

    @property
    def name(self) -> str:
        return self.get_name()['name']

    @name.setter
    def name(self, n: str) -> None:
        self.set_name({'name': n})

    @property
    def mode(self) -> str:
        return self.get_mode()['mode']

    @mode.setter
    def mode(self, m: str) -> None:
        self.set_mode({'mode': m})

    @property
    def version(self) -> str:
        return self.get_firmware_version()['version']

    def get_name(self):
        response = self._get('device_name')
        response.raise_for_status()
        return response.json()

    def set_name(self, data):
        response = self._post('device_name', json=data)
        response.raise_for_status()
        return response.json()

    def reset(self):
        response = self._get('reset')
        response.raise_for_status()
        return response.json()

    def get_network_status(self):
        response = self._get('network/status')
        response.raise_for_status()
        return response.json()

    def get_firmware_version(self):
        response = self._get('fw/version')
        response.raise_for_status()
        return response.json()

    def get_details(self):
        response = self._get('gestalt')
        response.raise_for_status()
        return response.json()

    def get_mode(self):
        response = self._get('led/mode')
        response.raise_for_status()
        return response.json()

    def set_mode(self, data):
        response = self._post('led/mode', json=data)
        response.raise_for_status()
        return response.json()

    def get_mqtt(self):
        response = self._get('mqtt/config')
        response.raise_for_status()
        return response.json()

    def set_mqtt(self, data):
        response = self._post('mqtt/config', json=data)
        response.raise_for_status()
        return response.json()

    def send_frame(self, frame: TWINKLY_FRAME):
        if len(frame) != self.length:
            raise ValueError("Invalid frame length")
        header = bytes([0x01]) + bytes(base64.b64decode(self.token)) + bytes([self.length])
        payload = []
        for x in frame:
            payload.extend(list(x))
        self.socket.sendto(header + bytes(payload), (self.host, self.rt_port))

    def get_movie_config(self):
        response = self._get('led/movie/config')
        response.raise_for_status()
        return response.json()

    def set_movie_config(self, data):
        response = self._post('led/movie/config', json=data)
        response.raise_for_status()
        return response.json()

    def upload_movie(self, movie: bytes):
        response = self._post('led/movie/full', data=movie,
                              headers={'Content-Type': 'application/octet-stream'})
        response.raise_for_status()
        return response.json()

    def realtime(self):
        self.mode = "rt"

    def movie(self):
        self.mode = "movie"

    def off(self):
        self.mode = "off"

    def demo(self):
        self.mode = "demo"


def main():
    """ Main function"""

    parser = argparse.ArgumentParser(description='Twinkly Twinkly Little Star')
    parser.add_argument('--host',
                        metavar='hostname',
                        required=True,
                        help='Device address')
    parser.add_argument('--debug',
                        action='store_true',
                        help="Enable debugging")
    parser.add_argument('--json',
                        action='store_true',
                        help="Output result as compact JSON")

    subparsers = parser.add_subparsers(dest='command')

    subparsers.add_parser('network', help="Get network status")
    subparsers.add_parser('firmware', help="Get firmware version")
    subparsers.add_parser('details', help="Get device details")
    subparsers.add_parser('token', help="Get authentication token")

    parser_name = subparsers.add_parser('name', help="Get or set device name")
    parser_name.add_argument('--name', metavar='name', type=str, required=False)

    parser_mode = subparsers.add_parser('mode', help="Get or set LED operation mode")
    parser_mode.add_argument('--mode', choices=TWINKLY_MODES, required=False)

    parser_mqtt = subparsers.add_parser('mqtt', help="Get or set MQTT configuration")
    parser_mqtt.add_argument('--json',
                             dest='mqtt_json',
                             metavar='mqtt',
                             type=str,
                             required=False,
                             help="MQTT config as JSON")

    parser_movie = subparsers.add_parser('movie', help="Movie configuration")
    parser_movie.add_argument('--delay',
                              dest='movie_delay',
                              metavar='milliseconds',
                              type=int,
                              default=100,
                              required=False,
                              help="Delay between frames")
    parser_movie.add_argument('--file',
                              dest='movie_file',
                              metavar='filename',
                              type=str,
                              required=True,
                              help="Movie file")

    args = parser.parse_args()

    if args.debug:
        logging.basicConfig(level=logging.DEBUG)
        requests_log = logging.getLogger("requests.packages.urllib3")
        requests_log.setLevel(logging.DEBUG)
        requests_log.propagate = True

    t = Twinkly(host=args.host)

    if args.command == 'name':
        res = t.get_name() if args.name is None else t.set_name({'name': args.name})
    elif args.command == 'network':
        res = t.get_network_status()
    elif args.command == 'firmware':
        res = t.get_firmware_version()
    elif args.command == 'details':
        res = t.get_details()
    elif args.command == 'mode':
        res = t.get_mode() if args.mode is None else t.set_mode({'mode': args.mode})
    elif args.command == 'mqtt':
        if args.mqtt_json is None:
            res = t.get_mqtt()
        else:
            data = json.loads(args.mqtt_json)
            res if args.mode is None else t.set_mqtt(data)
    elif args.command == 'movie':
        if args.movie_file:
            with open(args.movie_file, 'rb') as f:
                movie = f.read()
            params = {
                'frame_delay': args.movie_delay,
                'leds_number': t.length,
                'frames_number': int(len(movie) / 3 / t.length)
            }
            t.mode = 'movie'
            t.set_movie_config(params)
            res = t.upload_movie(movie)
        else:
            res = t.get_movie_config()
    else:
        raise Exception("Unknown command")

    if args.json:
        print(json.dumps(res, indent=None, separators=(',', ':')))
    else:
        print(json.dumps(res, indent=4))


if __name__ == "__main__":
    main()


ttls/colours.py
from dataclasses import dataclass
from typing import Dict, Optional, Tuple, Union

ColourDict = Dict[str, int]
ColourTuple = Union[Tuple[int, int, int], Tuple[int, int, int, int]]
TwinklyColourTuple = Union[Tuple[int, int, int], Tuple[int, int, int, int]]


@dataclass(frozen=True)
class TwinklyColour:
    red: int
    green: int
    blue: int
    white: Optional[int] = None

    def as_twinkly_tuple(self) -> TwinklyColourTuple:
        """Convert TwinklyColour to a tuple as used by Twinkly: (R,G,B) or (W,R,G,B)"""
        if self.white is not None:
            return (self.white, self.red, self.green, self.blue)
        else:
            return (self.red, self.green, self.blue)

    def as_tuple(self) -> ColourTuple:
        """Convert TwinklyColour to a tuple: (R,G,B) or (R,G,B,W)"""
        if self.white is not None:
            return (self.red, self.green, self.blue, self.white)
        else:
            return (self.red, self.green, self.blue)

    def __iter__(self):
        for i in self.as_tuple():
            yield i

    def as_dict(self) -> ColourDict:
        """Convert TwinklyColour to a dict wth color names used by set-led functions."""
        if self.white is not None:
            return {
                "red": self.red,
                "green": self.green,
                "blue": self.blue,
                "white": self.white,
            }
        else:
            return {"red": self.red, "green": self.green, "blue": self.blue}

    @classmethod
    def from_twinkly_tuple(cls, t):
        if len(t) == 4:
            return cls(red=t[1], green=t[2], blue=t[3], white=t[0])
        elif len(t) == 3:
            return cls(red=t[0], green=t[1], blue=t[2])
        raise TypeError("Unknown colour format")


ttls/client.py
"""
Twinkly Twinkly Little Star
https://github.com/jschlyter/ttls

Copyright (c) 2019 Jakob Schlyter. All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
"""

import base64
import logging
import os
import socket
import time
from itertools import cycle, islice
from typing import Any, Callable, Dict, List, Optional, Tuple, Union

from aiohttp import ClientResponseError, ClientSession, ClientTimeout
from aiohttp.web_exceptions import HTTPUnauthorized

from .colours import TwinklyColour, TwinklyColourTuple

_LOGGER = logging.getLogger(__name__)

TwinklyFrame = List[TwinklyColourTuple]
TwinklyResult = Optional[dict]


TWINKLY_MODES = [
    "color",
    "demo",
    "effect",
    "movie",
    "off",
    "playlist",
    "rt",
]
RT_PAYLOAD_MAX_LIGHTS = 300

TWINKLY_MUSIC_DRIVERS_OFFICIAL = {
    "VU Meter": "00000000-0000-0000-0000-000000000001",
    "Beat Hue": "00000000-0000-0000-0000-000000000002",
    "Psychedelica": "00000000-0000-0000-0000-000000000003",
    "Red Vertigo": "00000000-0000-0000-0000-000000000004",
    "Dancing Bands": "00000000-0000-0000-0000-000000000005",
    "Diamond Swirl": "00000000-0000-0000-0000-000000000006",
    "Joyful Stripes": "00000000-0000-0000-0000-000000000007",
    "Angel Fade": "00000000-0000-0000-0000-000000000008",
    "Clockwork": "00000000-0000-0000-0000-000000000009",
    "Sipario": "00000000-0000-0000-0000-00000000000A",
    "Sunset": "00000000-0000-0000-0000-00000000000B",
    "Elevator": "00000000-0000-0000-0000-00000000000C",
}

TWINKLY_MUSIC_DRIVERS_UNOFFICIAL = {
    "VU Meter 2": "00000000-0000-0000-0000-000001000001",
    "Beat Hue 2": "00000000-0000-0000-0000-000001000002",
    "Psychedelica 2": "00000000-0000-0000-0000-000001000003",
    "Sparkle": "00000000-0000-0000-0000-000001000005",
    "Sparkle Hue": "00000000-0000-0000-0000-000001000006",
    "Psycho Sparkle": "00000000-0000-0000-0000-000001000007",
    "Psycho Hue": "00000000-0000-0000-0000-000001000008",
    "Red Line": "00000000-0000-0000-0000-000001000009",
    "Red Vertigo 2": "00000000-0000-0000-0000-000002000004",
    "Dancing Bands 2": "00000000-0000-0000-0000-000002000005",
    "Diamond Swirl 2": "00000000-0000-0000-0000-000002000006",
    "Angel Fade 2": "00000000-0000-0000-0000-000002000008",
    "Clockwork 2": "00000000-0000-0000-0000-000002000009",
    "Sunset 2": "00000000-0000-0000-0000-00000200000B",
}

TWINKLY_MUSIC_DRIVERS = {
    **TWINKLY_MUSIC_DRIVERS_OFFICIAL,
    **TWINKLY_MUSIC_DRIVERS_UNOFFICIAL,
}

DEFAULT_TIMEOUT = 3


class Twinkly(object):
    def __init__(
        self,
        host: str,
        session: Optional[ClientSession] = None,
        timeout: Optional[int] = None,
    ):
        self.host = host
        self._timeout = ClientTimeout(total=timeout or DEFAULT_TIMEOUT)
        if session:
            self._session = session
            self._shared_session = True
        else:
            self._session = None
            self._shared_session = False
        self._socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self._headers: Dict[str, str] = {}
        self._rt_port = 7777
        self._expires = None
        self._token = None
        self._details: Dict[str, Union[str, int]] = {}
        self._default_mode = "movie"

    @property
    def base(self) -> str:
        return f"http://{self.host}/xled/v1"

    @property
    def length(self) -> int:
        return int(self._details["number_of_led"])

    @property
    def default_mode(self) -> str:
        return self._default_mode

    @default_mode.setter
    def default_mode(self, mode: str) -> str:
        if mode not in TWINKLY_MODES:
            raise ValueError("Invalid mode")
        if mode == "off":
            _LOGGER.warning("Setting default mode to off")
        self._default_mode = mode

    async def close(self) -> None:
        if not self._shared_session:
            await self._get_session().close()

    async def interview(self) -> None:
        if len(self._details) == 0:
            self._details = await self.get_details()

    def _get_session(self):
        return self._session or ClientSession()

    async def _post(self, endpoint: str, **kwargs) -> Any:
        await self.ensure_token()
        _LOGGER.debug("POST endpoint %s", endpoint)
        if "json" in kwargs:
            _LOGGER.debug("POST payload %s", kwargs["json"])
        headers = kwargs.pop("headers", self._headers)
        retry_num = kwargs.pop("retry_num", 0)
        try:
            async with self._get_session().post(
                f"{self.base}/{endpoint}",
                headers=headers,
                timeout=self._timeout,
                raise_for_status=True,
                **kwargs,
            ) as r:
                _LOGGER.debug("POST response %d", r.status)
                return await r.json()
        except ClientResponseError as e:
            if e.status == HTTPUnauthorized.status_code:
                await self._handle_authorized(
                    self._post, endpoint, exception=e, retry_num=retry_num, **kwargs
                )
            else:
                raise e

    async def _get(self, endpoint: str, **kwargs) -> Any:
        await self.ensure_token()
        _LOGGER.debug("GET endpoint %s", endpoint)
        headers = kwargs.pop("headers", self._headers)
        retry_num = kwargs.pop("retry_num", 0)
        try:
            async with self._get_session().get(
                f"{self.base}/{endpoint}",
                headers=headers,
                timeout=self._timeout,
                raise_for_status=True,
                **kwargs,
            ) as r:
                _LOGGER.debug("GET response %d", r.status)
                return await r.json()
        except ClientResponseError as e:
            if e.status == HTTPUnauthorized.status_code:
                return await self._handle_authorized(
                    self._get,
                    endpoint,
                    exception=e,
                    retry_num=retry_num,
                    **kwargs,
                )
            else:
                raise e

    async def _handle_authorized(
        self, request_method: Callable, endpoint: str, exception: Exception, **kwargs
    ) -> None:
        max_retries = 1
        retry_num = kwargs.pop("retry_num", 0)

        if retry_num >= max_retries:
            _LOGGER.debug(
                f"Invalid token for request. Maximum retries of {max_retries} exceeded."
            )
            raise exception

        retry_num += 1
        _LOGGER.debug(
            f"Invalid token for request. Refreshing token and attempting retry {retry_num} of {max_retries}."
        )
        await self.refresh_token()
        return await request_method(
            endpoint, headers=self._headers, retry_num=retry_num, **kwargs
        )

    async def refresh_token(self) -> None:
        await self.login()
        await self.verify_login()
        _LOGGER.debug("Authentication token refreshed")

    async def ensure_token(self) -> str:
        if self._expires is None or self._expires <= time.time():
            _LOGGER.debug("Authentication token expired, will refresh")
            await self.refresh_token()
        else:
            _LOGGER.debug("Authentication token still valid")
        return self._token or ""

    async def login(self) -> None:
        challenge = base64.b64encode(os.urandom(32)).decode()
        payload = {"challenge": challenge}
        async with self._get_session().post(
            f"{self.base}/login",
            json=payload,
            timeout=self._timeout,
            raise_for_status=True,
        ) as r:
            data = await r.json()
        self._token = data["authentication_token"]
        self._headers["X-Auth-Token"] = self._token
        self._expires = time.time() + data["authentication_token_expires_in"]

    async def logout(self) -> None:
        await self._post("logout", json={})
        self._token = None

    async def verify_login(self) -> None:
        await self._post("verify", json={})

    async def get_name(self) -> Any:
        return await self._get("device_name")

    async def set_name(self, name: str) -> Any:
        return await self._post("device_name", json={"name": name})

    async def reset(self) -> Any:
        return await self._get("reset")

    async def get_network_status(self) -> Any:
        return await self._get("network/status")

    async def get_firmware_version(self) -> Any:
        return await self._get("fw/version")

    async def get_details(self) -> Any:
        return await self._get("gestalt")

    async def is_on(self) -> Optional[bool]:
        mode = await self.get_mode()
        if mode is None:
            return None
        return mode.get("mode", "off") != "off"

    async def turn_on(self) -> Any:
        return await self.set_mode(self._default_mode)

    async def turn_off(self) -> Any:
        return await self.set_mode("off")

    async def get_brightness(self) -> Any:
        return await self._get("led/out/brightness")

    async def set_brightness(self, percent: int) -> Any:
        return await self._post(
            "led/out/brightness", json={"value": percent, "type": "A"}
        )

    async def get_mode(self) -> Any:
        return await self._get("led/mode")

    async def set_mode(self, mode: str) -> Any:
        return await self._post("led/mode", json={"mode": mode})

    async def get_mqtt(self) -> Any:
        return await self._get("mqtt/config")

    async def set_mqtt(self, data: dict) -> Any:
        return await self._post("mqtt/config", json=data)

    async def send_frame(self, frame: TwinklyFrame) -> None:
        await self.interview()
        if len(frame) != self.length:
            raise ValueError("Invalid frame length")
        token = await self.ensure_token()
        header = bytes([0x01]) + bytes(base64.b64decode(token)) + bytes([self.length])
        payload = []
        for x in frame:
            payload.extend(list(x))
        self._socket.sendto(header + bytes(payload), (self.host, self._rt_port))

    async def send_frame_2(self, frame: TwinklyFrame) -> None:
        await self.interview()
        if len(frame) != self.length:
            raise ValueError("Invalid frame length")
        token = await self.ensure_token()
        frame_segments = [
            frame[i : i + RT_PAYLOAD_MAX_LIGHTS]
            for i in range(0, len(frame), RT_PAYLOAD_MAX_LIGHTS)
        ]
        for i in range(0, len(frame_segments)):
            header = (
                bytes([len(frame_segments)])
                + bytes(base64.b64decode(token))
                + bytes([0, 0])
                + bytes([i])
            )
            payload = []
            for x in frame_segments[i]:
                payload.extend(list(x))
            self._socket.sendto(header + bytes(payload), (self.host, self._rt_port))

    async def get_movie_config(self) -> Any:
        return await self._get("led/movie/config")

    async def set_movie_config(self, data: dict) -> Any:
        return await self._post("led/movie/config", json=data)

    async def upload_movie(self, movie: bytes) -> Any:
        return await self._post(
            "led/movie/full",
            data=movie,
            headers={"Content-Type": "application/octet-stream"},
        )

    async def set_static_colour(
        self,
        colour: Union[
            TwinklyColour,
            TwinklyColourTuple,
            List[TwinklyColour],
            List[TwinklyColourTuple],
        ],
    ) -> None:
        if isinstance(colour, List):
            colour = colour[0]
        if isinstance(colour, Tuple):
            colour = TwinklyColour.from_twinkly_tuple(colour)
        await self._post(
            "led/color",
            json={"red": colour.red, "green": colour.green, "blue": colour.blue},
        )
        await self.set_mode("color")

    async def set_cycle_colours(
        self,
        colour: Union[
            TwinklyColour,
            TwinklyColourTuple,
            List[TwinklyColour],
            List[TwinklyColourTuple],
        ],
    ) -> None:
        if isinstance(colour, TwinklyColour):
            sequence = [colour.as_twinkly_tuple()]
        elif isinstance(colour, Tuple):
            sequence = [colour]
        elif isinstance(colour, List):
            if isinstance(colour[0], TwinklyColour):
                sequence = [c.as_twinkly_tuple() for c in color]
            else:
                sequence = colour
        else:
            raise TypeError("Unknown colour format")
        frame = list(islice(cycle(sequence), self.length))
        movie = bytes([item for t in frame for item in t])
        await self.upload_movie(movie)
        await self.set_movie_config(
            {
                "frames_number": 1,
                "loop_type": 0,
                "frame_delay": 1000,
                "leds_number": self.length,
            }
        )
        await self.set_mode("movie")

    async def summary(self) -> Any:
        return await self._get("summary")

    async def music_on(self) -> Any:
        return await self._post("music/enabled", json={"enabled": 1})

    async def music_off(self) -> Any:
        return await self._post("music/enabled", json={"enabled": 0})

    async def get_music_drivers(self) -> Any:
        """
        This endpoint is not currently used by the Twinkly app, but was discovered through
        trial & error.  It raises a 400 error ('unexpected content-length header') when
        called from aiohttp, but returns JSON when called via requests.
        This endpoint was used to map driver names to IDs and identify the 'unofficial'
        drivers on the device.
        {"code": 1000, "drivers_number": 26, "unique_ids": [<list of ID strings>]}
        """
        # return await self._get("music/drivers")
        raise NotImplementedError

    async def next_music_driver(self) -> Any:
        return await self._post("music/drivers/current", json={"action": "next"})

    async def previous_music_driver(self) -> Any:
        return await self._post("music/drivers/current", json={"action": "prev"})

    async def get_current_music_driver(self) -> Any:
        return await self._get("music/drivers/current")

    async def set_current_music_driver(self, driver_name: str) -> Any:
        unique_id = self._music_driver_id(driver_name)
        if not unique_id:
            _LOGGER.error(f"'{driver_name}' is an invalid music driver")
            return
        # An explicit driver cannot be set unless next/previous driver was called first
        current_driver = await self.get_current_music_driver()
        if current_driver["handle"] == -1:
            await self.next_music_driver()
        return await self._post("music/drivers/current", json={"unique_id": unique_id})

    def _music_driver_id(self, driver_name: str) -> Any:
        if driver_name in TWINKLY_MUSIC_DRIVERS_OFFICIAL:
            return TWINKLY_MUSIC_DRIVERS_OFFICIAL[driver_name]
        elif driver_name in TWINKLY_MUSIC_DRIVERS_UNOFFICIAL:
            _LOGGER.warn(
                f"Music driver '{driver_name}'is defined, but is not officially supported"
            )
            return TWINKLY_MUSIC_DRIVERS_UNOFFICIAL[driver_name]
        else:
            return None

    async def get_saved_movies(self) -> Any:
        return await self._get("movies")

    async def get_current_movie(self) -> Any:
        return await self._get("movies/current")

    async def set_current_movie(self, movie_id: int) -> Any:
        return await self._post("movies/current", json={"id": movie_id})

    async def get_current_colour(self) -> Any:
        return await self._get("led/color")

    async def get_predefined_effects(self) -> Any:
        """Get the list of predefined effects."""
        return await self._get("led/effects")

    async def get_current_predefined_effect(self) -> Any:
        """Get current effect."""
        return await self._get("led/effects/current")

    async def set_current_predefined_effect(self, effect_id: int) -> None:
        """Set current effect."""
        await self._post(
            "led/effects/current",
            json={"effect_id": effect_id},
        )

    async def get_playlist(self) -> Any:
        """Get the playlist."""
        return await self._get("playlist")

    async def get_current_playlist_entry(self) -> Any:
        """Get current playlist."""
        return await self._get("playlist/current")

    async def set_current_playlist_entry(self, entry_id: int) -> None:
        """Jump to specific effect in the playlist."""
        await self._post(
            "playlist/current",
            json={"id": entry_id},
        )
